File: ts/tests/store.test.ts
import { describe, it, expect } from "vitest";
import { createBootstrap, createEvent, NoopSigner } from "../src/event.js";
import { InMemoryStore } from "../src/store.js";
import { ActorId, ConversationId, EventId, EventType } from "../src/types.js";
import { ChainIntegrityError, EventNotFoundError } from "../src/errors.js";

const boot = () => createBootstrap(new ActorId("alice"), new NoopSigner());
const next = (prev: ReturnType<typeof boot>) => createEvent(
  new EventType("trust.updated"), new ActorId("alice"), {},
  [prev.id], new ConversationId("conv_1"), prev.hash, new NoopSigner(),
);

describe("InMemoryStore", () => {
  it("append and get", () => {
    const s = new InMemoryStore();
    const b = boot();
    s.append(b);
    expect(s.get(b.id).id.value).toBe(b.id.value);
  });

  it("head empty", () => expect(new InMemoryStore().head().isNone).toBe(true));

  it("head after append", () => {
    const s = new InMemoryStore();
    const b = boot();
    s.append(b);
    expect(s.head().unwrap().id.value).toBe(b.id.value);
  });

  it("count", () => {
    const s = new InMemoryStore();
    expect(s.count()).toBe(0);
    s.append(boot());
    expect(s.count()).toBe(1);
  });

  it("chain of events", () => {
    const s = new InMemoryStore();
    const b = boot(); s.append(b);
    const e1 = next(b); s.append(e1);
    const e2 = next(e1); s.append(e2);
    expect(s.count()).toBe(3);
  });

  it("rejects broken chain", () => {
    const s = new InMemoryStore();
    const b = boot(); s.append(b);
    const bad = createEvent(new EventType("trust.updated"), new ActorId("alice"), {}, [b.id], new ConversationId("c"), b.prevHash, new NoopSigner());
    expect(() => s.append(bad)).toThrow(ChainIntegrityError);
  });

  it("get nonexistent", () => {
    expect(() => new InMemoryStore().get(new EventId("019462a0-0000-7000-8000-000000000099"))).toThrow(EventNotFoundError);
  });

  it("verify chain valid", () => {
    const s = new InMemoryStore();
    const b = boot(); s.append(b); s.append(next(b));
    expect(s.verifyChain()).toEqual({ valid: true, length: 2 });
  });

  it("recent", () => {
    const s = new InMemoryStore();
    const b = boot(); s.append(b);
    const e1 = next(b); s.append(e1);
    const r = s.recent(2);
    expect(r).toHaveLength(2);
    expect(r[0].id.value).toBe(e1.id.value);
  });

  it("byType returns matching events newest first", () => {
    const s = new InMemoryStore();
    const b = boot(); s.append(b);
    const e1 = next(b); s.append(e1);
    const e2 = next(e1); s.append(e2);

    const results = s.byType(new EventType("trust.updated"), 10);
    expect(results).toHaveLength(2);
    expect(results[0].id.value).toBe(e2.id.value);
    expect(results[1].id.value).toBe(e1.id.value);
  });

  it("byType respects limit", () => {
    const s = new InMemoryStore();
    const b = boot(); s.append(b);
    const e1 = next(b); s.append(e1);
    const e2 = next(e1); s.append(e2);

    const results = s.byType(new EventType("trust.updated"), 1);
    expect(results).toHaveLength(1);
    expect(results[0].id.value).toBe(e2.id.value);
  });

  it("byType returns empty for no matches", () => {
    const s = new InMemoryStore();
    const b = boot(); s.append(b);
    expect(s.byType(new EventType("trust.updated"), 10)).toHaveLength(0);
  });

  it("bySource returns matching events newest first", () => {
    const s = new InMemoryStore();
    const b = boot(); s.append(b);
    const e1 = next(b); s.append(e1);

    const results = s.bySource(new ActorId("alice"), 10);
    expect(results).toHaveLength(2);
    expect(results[0].id.value).toBe(e1.id.value);
    expect(results[1].id.value).toBe(b.id.value);
  });

  it("bySource returns empty for unknown actor", () => {
    const s = new InMemoryStore();
    const b = boot(); s.append(b);
    expect(s.bySource(new ActorId("bob"), 10)).toHaveLength(0);
  });

  it("byConversation returns matching events newest first", () => {
    const s = new InMemoryStore();
    const b = boot(); s.append(b);
    const e1 = next(b); s.append(e1);

    const results = s.byConversation(new ConversationId("conv_1"), 10);
    expect(results).toHaveLength(1);
    expect(results[0].id.value).toBe(e1.id.value);
  });

  it("byConversation returns empty for unknown conversation", () => {
    const s = new InMemoryStore();
    const b = boot(); s.append(b);
    expect(s.byConversation(new ConversationId("conv_unknown"), 10)).toHaveLength(0);
  });

  it("ancestors returns causal ancestors", () => {
    const s = new InMemoryStore();
    const b = boot(); s.append(b);
    const e1 = next(b); s.append(e1);
    const e2 = next(e1); s.append(e2);

    const anc = s.ancestors(e2.id, 10);
    expect(anc).toHaveLength(2);
    // e1 is direct cause, b is cause of e1
    const ids = anc.map((e) => e.id.value);
    expect(ids).toContain(e1.id.value);
    expect(ids).toContain(b.id.value);
  });

  it("ancestors respects maxDepth", () => {
    const s = new InMemoryStore();
    const b = boot(); s.append(b);
    const e1 = next(b); s.append(e1);
    const e2 = next(e1); s.append(e2);

    const anc = s.ancestors(e2.id, 1);
    expect(anc).toHaveLength(1);
    expect(anc[0].id.value).toBe(e1.id.value);
  });

  it("ancestors throws for unknown event", () => {
    const s = new InMemoryStore();
    expect(() => s.ancestors(new EventId("019462a0-0000-7000-8000-000000000099"), 10)).toThrow(EventNotFoundError);
  });

  it("descendants returns causal descendants", () => {
    const s = new InMemoryStore();
    const b = boot(); s.append(b);
    const e1 = next(b); s.append(e1);
    const e2 = next(e1); s.append(e2);

    const desc = s.descendants(b.id, 10);
    // e1 has b in causes, e2 has e1 in causes
    const ids = desc.map((e) => e.id.value);
    expect(ids).toContain(e1.id.value);
    expect(ids).toContain(e2.id.value);
  });

  it("descendants respects maxDepth", () => {
    const s = new InMemoryStore();
    const b = boot(); s.append(b);
    const e1 = next(b); s.append(e1);
    const e2 = next(e1); s.append(e2);

    const desc = s.descendants(b.id, 1);
    expect(desc).toHaveLength(1);
    expect(desc[0].id.value).toBe(e1.id.value);
  });

  it("descendants throws for unknown event", () => {
    const s = new InMemoryStore();
    expect(() => s.descendants(new EventId("019462a0-0000-7000-8000-000000000099"), 10)).toThrow(EventNotFoundError);
  });
});

File: rust/tests/store_test.rs
use eventgraph::event::*;
use eventgraph::store::*;
use eventgraph::types::*;

fn bootstrap() -> Event {
    create_bootstrap(ActorId::new("alice").unwrap(), &NoopSigner, 1)
}

#[test]
fn append_and_get() {
    let mut store = InMemoryStore::new();
    let ev = bootstrap();
    let id = ev.id.clone();
    store.append(ev).unwrap();
    assert_eq!(store.get(&id).unwrap().id, id);
}

#[test]
fn head_returns_latest() {
    let mut store = InMemoryStore::new();
    assert!(store.head().is_none());
    let ev = bootstrap();
    store.append(ev).unwrap();
    assert!(store.head().is_some());
}

#[test]
fn count_increments() {
    let mut store = InMemoryStore::new();
    assert_eq!(store.count(), 0);
    store.append(bootstrap()).unwrap();
    assert_eq!(store.count(), 1);
}

#[test]
fn chain_integrity_enforced() {
    let mut store = InMemoryStore::new();
    let boot = bootstrap();
    store.append(boot).unwrap();

    // Create a second bootstrap (wrong prev_hash)
    let bad = bootstrap();
    assert!(store.append(bad).is_err());
}

#[test]
fn verify_chain_valid() {
    let mut store = InMemoryStore::new();
    let boot = bootstrap();
    store.append(boot).unwrap();
    let v = store.verify_chain();
    assert!(v.valid);
    assert_eq!(v.length, 1);
}

#[test]
fn get_nonexistent_fails() {
    let store = InMemoryStore::new();
    let id = EventId::new("019462a0-0000-7000-8000-000000000001").unwrap();
    assert!(store.get(&id).is_err());
}

#[test]
fn chained_append() {
    let signer = NoopSigner;
    let mut store = InMemoryStore::new();
    let boot = bootstrap();
    let prev_hash = boot.hash.clone();
    let boot_id = boot.id.clone();
    store.append(boot).unwrap();

    let mut content = std::collections::BTreeMap::new();
    content.insert("score".to_string(), serde_json::Value::Number(serde_json::Number::from_f64(0.5).unwrap()));

    let ev = create_event(
        EventType::new("trust.updated").unwrap(),
        ActorId::new("alice").unwrap(),
        content,
        vec![boot_id],
        ConversationId::new("conv_1").unwrap(),
        prev_hash,
        &signer,
        1,
    );
    store.append(ev).unwrap();
    assert_eq!(store.count(), 2);
    let v = store.verify_chain();
    assert!(v.valid);
    assert_eq!(v.length, 2);
}

#[test]
fn recent_returns_reverse_order() {
    let mut store = InMemoryStore::new();
    store.append(bootstrap()).unwrap();
    let recent = store.recent(10);
    assert_eq!(recent.len(), 1);
}

// ── Helper to build a chained event ────────────────────────────────────

fn append_event(
    store: &mut InMemoryStore,
    event_type: &str,
    source: &str,
    conversation_id: &str,
    causes: Vec<EventId>,
) -> Event {
    let head = store.head().unwrap();
    let prev_hash = head.hash.clone();
    let ev = create_event(
        EventType::new(event_type).unwrap(),
        ActorId::new(source).unwrap(),
        std::collections::BTreeMap::new(),
        causes,
        ConversationId::new(conversation_id).unwrap(),
        prev_hash,
        &NoopSigner,
        1,
    );
    store.append(ev).unwrap()
}

// ── by_type ────────────────────────────────────────────────────────────

#[test]
fn by_type_filters_correctly() {
    let mut store = InMemoryStore::new();
    let boot = bootstrap();
    let boot_id = boot.id.clone();
    store.append(boot).unwrap();

    let ev1 = append_event(&mut store, "trust.updated", "alice", "conv_1", vec![boot_id.clone()]);
    let ev1_id = ev1.id.clone();
    append_event(&mut store, "message.sent", "alice", "conv_1", vec![ev1_id.clone()]);
    append_event(&mut store, "trust.updated", "alice", "conv_1", vec![ev1_id]);

    let trust_type = EventType::new("trust.updated").unwrap();
    let results = store.by_type(&trust_type, 10);
    assert_eq!(results.len(), 2);
    // Newest first
    assert_eq!(results[0].event_type.value(), "trust.updated");
    assert_eq!(results[1].event_type.value(), "trust.updated");
}

#[test]
fn by_type_respects_limit() {
    let mut store = InMemoryStore::new();
    let boot = bootstrap();
    let boot_id = boot.id.clone();
    store.append(boot).unwrap();

    let ev1 = append_event(&mut store, "trust.updated", "alice", "conv_1", vec![boot_id]);
    append_event(&mut store, "trust.updated", "alice", "conv_1", vec![ev1.id.clone()]);

    let trust_type = EventType::new("trust.updated").unwrap();
    let results = store.by_type(&trust_type, 1);
    assert_eq!(results.len(), 1);
}

#[test]
fn by_type_returns_empty_for_no_match() {
    let mut store = InMemoryStore::new();
    store.append(bootstrap()).unwrap();

    let t = EventType::new("nonexistent.type").unwrap();
    let results = store.by_type(&t, 10);
    assert!(results.is_empty());
}

// ── by_source ──────────────────────────────────────────────────────────

#[test]
fn by_source_filters_correctly() {
    let mut store = InMemoryStore::new();
    let boot = bootstrap(); // source = alice
    let boot_id = boot.id.clone();
    store.append(boot).unwrap();

    let ev1 = append_event(&mut store, "trust.updated", "bob", "conv_1", vec![boot_id.clone()]);
    append_event(&mut store, "trust.updated", "alice", "conv_1", vec![ev1.id.clone()]);

    let bob = ActorId::new("bob").unwrap();
    let results = store.by_source(&bob, 10);
    assert_eq!(results.len(), 1);
    assert_eq!(results[0].source.value(), "bob");
}

#[test]
fn by_source_respects_limit() {
    let mut store = InMemoryStore::new();
    let boot = bootstrap();
    let boot_id = boot.id.clone();
    store.append(boot).unwrap();

    let ev1 = append_event(&mut store, "trust.updated", "alice", "conv_1", vec![boot_id]);
    append_event(&mut store, "trust.updated", "alice", "conv_1", vec![ev1.id.clone()]);

    let alice = ActorId::new("alice").unwrap();
    // bootstrap + 2 events = 3 from alice, but limit to 2
    let results = store.by_source(&alice, 2);
    assert_eq!(results.len(), 2);
}

// ── by_conversation ────────────────────────────────────────────────────

#[test]
fn by_conversation_filters_correctly() {
    let mut store = InMemoryStore::new();
    let boot = bootstrap(); // conversation = conv_alice
    let boot_id = boot.id.clone();
    store.append(boot).unwrap();

    let ev1 = append_event(&mut store, "trust.updated", "alice", "conv_a", vec![boot_id.clone()]);
    append_event(&mut store, "trust.updated", "alice", "conv_b", vec![ev1.id.clone()]);

    let conv_a = ConversationId::new("conv_a").unwrap();
    let results = store.by_conversation(&conv_a, 10);
    assert_eq!(results.len(), 1);
    assert_eq!(results[0].conversation_id.value(), "conv_a");
}

#[test]
fn by_conversation_respects_limit() {
    let mut store = InMemoryStore::new();
    let boot = bootstrap();
    let boot_id = boot.id.clone();
    store.append(boot).unwrap();

    let ev1 = append_event(&mut store, "trust.updated", "alice", "conv_a", vec![boot_id]);
    append_event(&mut store, "trust.updated", "alice", "conv_a", vec![ev1.id.clone()]);

    let conv_a = ConversationId::new("conv_a").unwrap();
    let results = store.by_conversation(&conv_a, 1);
    assert_eq!(results.len(), 1);
}

// ── ancestors ──────────────────────────────────────────────────────────

#[test]
fn ancestors_returns_causal_parents() {
    let mut store = InMemoryStore::new();
    let boot = bootstrap();
    let boot_id = boot.id.clone();
    store.append(boot).unwrap();

    let ev1 = append_event(&mut store, "trust.updated", "alice", "conv_1", vec![boot_id.clone()]);
    let ev2 = append_event(&mut store, "trust.updated", "alice", "conv_1", vec![ev1.id.clone()]);

    // ancestors of ev2 at depth 1 should be ev1
    let ancestors = store.ancestors(&ev2.id, 1).unwrap();
    assert_eq!(ancestors.len(), 1);
    assert_eq!(ancestors[0].id, ev1.id);

    // ancestors of ev2 at depth 2 should be ev1 + boot
    let ancestors = store.ancestors(&ev2.id, 2).unwrap();
    assert_eq!(ancestors.len(), 2);
}

#[test]
fn ancestors_error_on_not_found() {
    let store = InMemoryStore::new();
    let id = EventId::new("019462a0-0000-7000-8000-000000000001").unwrap();
    assert!(store.ancestors(&id, 5).is_err());
}

#[test]
fn ancestors_depth_zero_returns_empty() {
    let mut store = InMemoryStore::new();
    let boot = bootstrap();
    let boot_id = boot.id.clone();
    store.append(boot).unwrap();

    let ancestors = store.ancestors(&boot_id, 0).unwrap();
    assert!(ancestors.is_empty());
}

// ── descendants ────────────────────────────────────────────────────────

#[test]
fn descendants_returns_causal_children() {
    let mut store = InMemoryStore::new();
    let boot = bootstrap();
    let boot_id = boot.id.clone();
    store.append(boot).unwrap();

    let ev1 = append_event(&mut store, "trust.updated", "alice", "conv_1", vec![boot_id.clone()]);
    append_event(&mut store, "trust.updated", "alice", "conv_1", vec![ev1.id.clone()]);

    // descendants of boot at depth 1 should include ev1 (ev1 references boot in causes)
    let desc = store.descendants(&boot_id, 1).unwrap();
    assert_eq!(desc.len(), 1);
    assert_eq!(desc[0].id, ev1.id);

    // descendants of boot at depth 2 should include ev1 and ev2
    let desc = store.descendants(&boot_id, 2).unwrap();
    assert_eq!(desc.len(), 2);
}

#[test]
fn descendants_error_on_not_found() {
    let store = InMemoryStore::new();
    let id = EventId::new("019462a0-0000-7000-8000-000000000001").unwrap();
    assert!(store.descendants(&id, 5).is_err());
}

#[test]
fn descendants_depth_zero_returns_empty() {
    let mut store = InMemoryStore::new();
    let boot = bootstrap();
    let boot_id = boot.id.clone();
    store.append(boot).unwrap();

    let desc = store.descendants(&boot_id, 0).unwrap();
    assert!(desc.is_empty());
}
